File: 2019/01_the_tyranny_of_the_rocket_equation/aoc01.py
"""The Tyranny of the Rocket Equation

Advent of Code 2019, day 1
Solution by Geir Arne Hjelle, 2019-12-01
"""
import sys


def fuel(mass):
    """Calculate fuel based on mass"""
    return max(mass // 3 - 2, 0)


def recursive_fuel(mass):
    """Calculate fuel based on mass, include fuel to cover mass of fuel"""
    return 0 if (fuel := mass // 3 - 2) <= 0 else fuel + recursive_fuel(fuel)


def main():
    for filename in sys.argv[1:]:
        if filename.startswith("--"):
            continue

        print(f"\n{filename}:")

        # Part 1
        with open(filename, mode="r") as fid:
            total_fuel = sum(fuel(int(line)) for line in fid)
        print(f"Total fuel:  {total_fuel}")

        # Part 2
        with open(filename, mode="r") as fid:
            rec_fuel = sum(recursive_fuel(int(line)) for line in fid)
        print(f"Total fuel, including fuel for fuel:  {rec_fuel}")


if __name__ == "__main__":
    debug = print if "--debug" in sys.argv else lambda *_: None
    main()

File: 2019/01_the_tyranny_of_the_rocket_equation/aoc01.jl
# The Tyranny of the Rocket Equation
#
# Advent of Code 2019, day 1
# Solution by Geir Arne Hjelle, 2020-11-30

using Pipe

# Calculate fuel based on mass
function fuel(mass)
    floor(mass / 3) - 2 |> Int64
end


# Calculate fuel based on mass, include fuel to cover mass of fuel
function recursive_fuel(mass)
    fuel_for_mass = mass |> fuel

    if fuel_for_mass <= 0
        return 0
    else
        return fuel_for_mass + recursive_fuel(fuel_for_mass)
    end
end


# Solve the problem for one file
function solve(filename)
    println("\n$(filename)")

    # Read from file
    input = open(filename) do fid
        @pipe fid |> readlines .|> parse(Int64, _)
    end

    # Part 1
    input .|> fuel |> sum |> println

    # Part 2
    input .|> recursive_fuel |> sum |> println
end


# Run solve on each file
ARGS .|> solve